// @openfate/bazi-engine/dist/core/cycles.js
"use strict";
// ============================================================================
// @openfate/bazi-engine — Da Yun (Major 10-Year Luck Cycles) Calculator
// ============================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateDaYun = calculateDaYun;
const pillars_1 = require("./pillars");
const tenGods_1 = require("./tenGods");
/**
 * calculateDaYun
 *
 * Calculates the nine 10-year Major Luck Cycles (Da Yun) from the natal EightChar.
 * Direction (forward/backward) is determined by gender and the Yin/Yang of the year stem:
 *   - Yang Year + Male   → Forward (Shun)
 *   - Yang Year + Female → Backward (Ni)
 *   - Yin Year  + Male   → Backward (Ni)
 *   - Yin Year  + Female → Forward (Shun)
 *
 * @param eightChar - The EightChar object from lunar-javascript (returned by generatePillarsFromSolar)
 * @param gender    - 'male' or 'female'
 * @param birthYear - Solar birth year (used to compute startAge)
 */
function calculateDaYun(eightChar, gender, birthYear) {
    const genderInt = gender === 'male' ? 1 : 0;
    const yun = eightChar.getYun(genderInt);
    const dayStem = eightChar.getDayGan();
    const rawCycles = yun.getDaYun();
    const cycles = [];
    // Index 0 is the natal chart segment — cycles start at index 1
    const MAX_CYCLES = 9;
    for (let i = 1; i <= MAX_CYCLES && i < rawCycles.length; i++) {
        const dy = rawCycles[i];
        const ganZhi = dy.getGanZhi();
        const stem = ganZhi.substring(0, 1);
        const branch = ganZhi.substring(1, 2);
        const startYear = dy.getStartYear();
        const endYear = dy.getEndYear();
        cycles.push({
            index: i,
            stem,
            branch,
            ganZhi,
            startYear,
            startAge: startYear - birthYear,
            endYear,
            endAge: endYear - birthYear,
            stemTenGod: (0, tenGods_1.calculateTenGod)(dayStem, stem),
            branchTenGod: (0, tenGods_1.calculateTenGod)(dayStem, (0, pillars_1.getMainQi)(branch)),
        });
    }
    const firstCycle = cycles[0];
    if (!firstCycle)
        throw new Error('Unable to calculate Da Yun cycles.');
    return {
        cycles,
        isForward: yun.isForward(),
        startYear: firstCycle.startYear,
        startAge: firstCycle.startAge,
        startDate: yun.getStartSolar().toYmdHms(),
        startOffset: {
            years: yun.getStartYear(),
            months: yun.getStartMonth(),
            days: yun.getStartDay(),
            hours: yun.getStartHour(),
        },
    };
}
