// @openfate/bazi-engine/dist/core/pillars.js
"use strict";
// ============================================================================
// @openfate/bazi-engine — Four Pillars Generator
// ============================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.generatePillarsFromSolar = generatePillarsFromSolar;
exports.getMainQi = getMainQi;
exports.getStemInfo = getStemInfo;
const lunar_javascript_1 = require("lunar-javascript");
const constants_1 = require("../constants");
function getRequiredElement(value, mapping, label) {
    const result = mapping[value];
    if (!result)
        throw new Error(`Unsupported ${label}: ${value}`);
    return result;
}
function getRequiredPolarity(stem) {
    const polarity = constants_1.STEM_TO_POLARITY[stem];
    if (!polarity)
        throw new Error(`Unsupported Heavenly Stem: ${stem}`);
    return polarity;
}
function buildPillar(stem, branch, details) {
    const hiddenStemDefinitions = constants_1.BRANCH_HIDDEN_STEMS[branch];
    if (!hiddenStemDefinitions)
        throw new Error(`Unsupported Earthly Branch: ${branch}`);
    return {
        stem,
        branch,
        element: getRequiredElement(stem, constants_1.STEM_TO_ELEMENT, 'Heavenly Stem'),
        ganZhi: `${stem}${branch}`,
        stemPolarity: getRequiredPolarity(stem),
        stemTenGod: details.stemTenGod,
        branchElement: getRequiredElement(branch, constants_1.BRANCH_TO_ELEMENT, 'Earthly Branch'),
        hiddenStems: hiddenStemDefinitions.map((hiddenStem, index) => ({
            stem: hiddenStem.stem,
            element: getRequiredElement(hiddenStem.stem, constants_1.STEM_TO_ELEMENT, 'Hidden Stem'),
            polarity: getRequiredPolarity(hiddenStem.stem),
            tenGod: details.hiddenStemTenGods[index],
            isMain: Boolean(hiddenStem.isMain),
        })),
        naYin: details.naYin,
        xun: details.xun,
        voidBranches: details.voidBranches.split(''),
        growthStage: details.growthStage,
    };
}
/**
 * generatePillarsFromSolar
 *
 * Converts a solar datetime (post True Solar Time correction) into the Four Pillars.
 * Uses lunar-javascript for Solar Term (节气) boundary detection and pillar generation.
 *
 * @param year  - Solar year (after TST correction)
 * @param month - Solar month
 * @param day   - Solar day
 * @param hour  - Hour (0-23), undefined if birth time unknown
 * @param minute - Minute
 * @param second - Second
 * @param dayBoundaryMode - MIDNIGHT_00 (default) or ZI_HOUR_23 (traditional)
 */
function generatePillarsFromSolar(year, month, day, hour, minute, second, dayBoundaryMode = 'MIDNIGHT_00') {
    const hasTime = hour !== undefined && hour !== null;
    const h = hasTime ? hour : 0;
    const m = minute !== null && minute !== void 0 ? minute : 0;
    const s = second !== null && second !== void 0 ? second : 0;
    const solar = hasTime
        ? lunar_javascript_1.Solar.fromYmdHms(year, month, day, h, m, s)
        : lunar_javascript_1.Solar.fromYmd(year, month, day);
    const lunar = solar.getLunar();
    const eightChar = lunar.getEightChar();
    // Sect 1 = Day shifts at 23:00 (traditional Zi Hour)
    // Sect 2 = Day shifts at 00:00 (midnight, modern default)
    eightChar.setSect(dayBoundaryMode === 'ZI_HOUR_23' ? 1 : 2);
    const dayStem = eightChar.getDayGan();
    return {
        pillars: {
            year: buildPillar(eightChar.getYearGan(), eightChar.getYearZhi(), {
                stemTenGod: eightChar.getYearShiShenGan(),
                hiddenStemTenGods: eightChar.getYearShiShenZhi(),
                naYin: eightChar.getYearNaYin(),
                xun: eightChar.getYearXun(),
                voidBranches: eightChar.getYearXunKong(),
                growthStage: eightChar.getYearDiShi(),
            }),
            month: buildPillar(eightChar.getMonthGan(), eightChar.getMonthZhi(), {
                stemTenGod: eightChar.getMonthShiShenGan(),
                hiddenStemTenGods: eightChar.getMonthShiShenZhi(),
                naYin: eightChar.getMonthNaYin(),
                xun: eightChar.getMonthXun(),
                voidBranches: eightChar.getMonthXunKong(),
                growthStage: eightChar.getMonthDiShi(),
            }),
            day: buildPillar(dayStem, eightChar.getDayZhi(), {
                stemTenGod: eightChar.getDayShiShenGan(),
                hiddenStemTenGods: eightChar.getDayShiShenZhi(),
                naYin: eightChar.getDayNaYin(),
                xun: eightChar.getDayXun(),
                voidBranches: eightChar.getDayXunKong(),
                growthStage: eightChar.getDayDiShi(),
            }),
            hour: hasTime ? buildPillar(eightChar.getTimeGan(), eightChar.getTimeZhi(), {
                stemTenGod: eightChar.getTimeShiShenGan(),
                hiddenStemTenGods: eightChar.getTimeShiShenZhi(),
                naYin: eightChar.getTimeNaYin(),
                xun: eightChar.getTimeXun(),
                voidBranches: eightChar.getTimeXunKong(),
                growthStage: eightChar.getTimeDiShi(),
            }) : null,
        },
        eightChar,
        dayStem,
    };
}
/**
 * getMainQi - Returns the main hidden stem of a branch
 */
function getMainQi(branch) {
    var _a, _b, _c;
    const hidden = (_a = constants_1.BRANCH_HIDDEN_STEMS[branch]) !== null && _a !== void 0 ? _a : [];
    const main = (_b = hidden.find(h => h.isMain)) !== null && _b !== void 0 ? _b : hidden[0];
    return (_c = main === null || main === void 0 ? void 0 : main.stem) !== null && _c !== void 0 ? _c : '';
}
/**
 * getStemInfo - Returns enriched stem details
 */
function getStemInfo(stem) {
    return {
        char: stem,
        pinyin: constants_1.STEM_TO_PINYIN[stem],
        element: getRequiredElement(stem, constants_1.STEM_TO_ELEMENT, 'Heavenly Stem'),
        polarity: getRequiredPolarity(stem),
    };
}
