// @openfate/bazi-engine/dist/core/tenGods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateTenGod = calculateTenGod;
const constants_1 = require("../constants");
const ELEMENTS = ['wood', 'fire', 'earth', 'metal', 'water'];
const TEN_GOD_NAMES = {
    self: { same: '比肩', different: '劫财' },
    output: { same: '食神', different: '伤官' },
    wealth: { same: '偏财', different: '正财' },
    power: { same: '七杀', different: '正官' },
    resource: { same: '偏印', different: '正印' },
};
function getRelationship(dayMaster, target) {
    const dayMasterIndex = ELEMENTS.indexOf(dayMaster);
    const targetIndex = ELEMENTS.indexOf(target);
    const difference = (targetIndex - dayMasterIndex + ELEMENTS.length) % ELEMENTS.length;
    if (difference === 0)
        return 'self';
    if (difference === 1)
        return 'output';
    if (difference === 2)
        return 'wealth';
    if (difference === 3)
        return 'power';
    return 'resource';
}
/**
 * calculateTenGod - Returns the classical Ten God relationship between two stems.
 */
function calculateTenGod(dayMaster, targetStem) {
    const dayMasterElement = constants_1.STEM_TO_ELEMENT[dayMaster];
    const targetElement = constants_1.STEM_TO_ELEMENT[targetStem];
    const dayMasterPolarity = constants_1.STEM_TO_POLARITY[dayMaster];
    const targetPolarity = constants_1.STEM_TO_POLARITY[targetStem];
    if (!dayMasterElement || !dayMasterPolarity)
        throw new Error(`Unsupported Day Master: ${dayMaster}`);
    if (!targetElement || !targetPolarity)
        throw new Error(`Unsupported target stem: ${targetStem}`);
    const relationship = getRelationship(dayMasterElement, targetElement);
    const names = TEN_GOD_NAMES[relationship];
    return dayMasterPolarity === targetPolarity ? names.same : names.different;
}
