// @openfate/bazi-engine/dist/index.js
"use strict";
// ============================================================================
// @openfate/bazi-engine — Public Entry Point
// ============================================================================
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateBaziInput = exports.BaziInputError = exports.calculateTenGod = exports.calculateDaYun = exports.getMainQi = exports.getStemInfo = exports.generatePillarsFromSolar = exports.detectInteractions = void 0;
exports.calculateBaziChart = calculateBaziChart;
const true_solar_time_1 = require("@openfate/true-solar-time");
const lunar_javascript_1 = require("lunar-javascript");
const pillars_1 = require("./core/pillars");
const cycles_1 = require("./core/cycles");
const interactions_1 = require("./core/interactions");
const validation_1 = require("./core/validation");
// Re-export all public types and constants
__exportStar(require("./types"), exports);
__exportStar(require("./constants"), exports);
var interactions_2 = require("./core/interactions");
Object.defineProperty(exports, "detectInteractions", { enumerable: true, get: function () { return interactions_2.detectInteractions; } });
var pillars_2 = require("./core/pillars");
Object.defineProperty(exports, "generatePillarsFromSolar", { enumerable: true, get: function () { return pillars_2.generatePillarsFromSolar; } });
Object.defineProperty(exports, "getStemInfo", { enumerable: true, get: function () { return pillars_2.getStemInfo; } });
Object.defineProperty(exports, "getMainQi", { enumerable: true, get: function () { return pillars_2.getMainQi; } });
var cycles_2 = require("./core/cycles");
Object.defineProperty(exports, "calculateDaYun", { enumerable: true, get: function () { return cycles_2.calculateDaYun; } });
var tenGods_1 = require("./core/tenGods");
Object.defineProperty(exports, "calculateTenGod", { enumerable: true, get: function () { return tenGods_1.calculateTenGod; } });
var validation_2 = require("./core/validation");
Object.defineProperty(exports, "BaziInputError", { enumerable: true, get: function () { return validation_2.BaziInputError; } });
Object.defineProperty(exports, "validateBaziInput", { enumerable: true, get: function () { return validation_2.validateBaziInput; } });
function createDateTime(year, month, day, hour, minute, second) {
    const hasTime = hour !== undefined;
    return {
        year,
        month,
        day,
        hour: hasTime ? hour : null,
        minute: hasTime ? minute !== null && minute !== void 0 ? minute : 0 : null,
        second: hasTime ? second !== null && second !== void 0 ? second : 0 : null,
    };
}
function createLunarDateTime(solarDateTime) {
    const solar = solarDateTime.hour === null
        ? lunar_javascript_1.Solar.fromYmd(solarDateTime.year, solarDateTime.month, solarDateTime.day)
        : lunar_javascript_1.Solar.fromYmdHms(solarDateTime.year, solarDateTime.month, solarDateTime.day, solarDateTime.hour, solarDateTime.minute, solarDateTime.second);
    const lunar = solar.getLunar();
    const lunarMonth = lunar.getMonth();
    return {
        lunar: {
            year: lunar.getYear(),
            month: Math.abs(lunarMonth),
            day: lunar.getDay(),
            hour: solarDateTime.hour,
            minute: solarDateTime.minute,
            second: solarDateTime.second,
            isLeapMonth: lunarMonth < 0,
        },
        zodiac: lunar.getYearShengXiao(),
    };
}
/**
 * calculateBaziChart
 *
 * The main entry point. Takes a birth configuration and returns a fully
 * calculated Bazi chart with Four Pillars, Da Yun cycles, and branch interactions.
 *
 * True Solar Time correction is applied automatically if longitude is provided.
 *
 * @example
 * ```typescript
 * const chart = calculateBaziChart({
 *   year: 1998, month: 12, day: 13, hour: 12, minute: 0,
 *   gender: 'female', longitude: 116.39, timezone: 8,
 * });
 *
 * console.log(chart.pillars.year.stem);   // '戊'
 * console.log(chart.pillars.year.branch); // '寅'
 * console.log(chart.daYun.cycles[0].ganZhi); // First major cycle
 * ```
 */
function calculateBaziChart(input) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
    (0, validation_1.validateBaziInput)(input);
    let solarYear = input.year;
    let solarMonth = input.month;
    let solarDay = input.day;
    let solarHour = input.hour;
    let solarMinute = (_a = input.minute) !== null && _a !== void 0 ? _a : 0;
    let solarSecond = 0;
    let solarTimeInfo = null;
    const inputType = (_b = input.calendarType) !== null && _b !== void 0 ? _b : 'solar';
    // ── 1. Lunar → Solar Calendar Conversion ────────────────────────────────
    if (input.calendarType === 'lunar') {
        // lunar-javascript: negative month represents a leap month
        const lunarMonth = input.isLeapMonth ? -input.month : input.month;
        const lunar = lunar_javascript_1.Lunar.fromYmd(input.year, lunarMonth, input.day);
        const solar = lunar.getSolar();
        solarYear = solar.getYear();
        solarMonth = solar.getMonth();
        solarDay = solar.getDay();
    }
    const civilSolar = createDateTime(solarYear, solarMonth, solarDay, input.hour, input.hour === undefined ? undefined : solarMinute, input.hour === undefined ? undefined : solarSecond);
    // ── 2. True Solar Time Correction ───────────────────────────────────────
    const enableTST = ((_c = input.enableTrueSolarTime) !== null && _c !== void 0 ? _c : true) &&
        input.longitude !== undefined && input.longitude !== null &&
        input.hour !== undefined;
    if (enableTST) {
        const civilTimeInput = input.timezone !== undefined
            ? {
                year: solarYear,
                month: solarMonth,
                day: solarDay,
                hour: solarHour,
                minute: solarMinute,
                timeZoneOffset: input.timezone,
                dstOffset: (_d = input.dstOffset) !== null && _d !== void 0 ? _d : 0,
            }
            : {
                year: solarYear,
                month: solarMonth,
                day: solarDay,
                hour: solarHour,
                minute: solarMinute,
                timeZoneId: input.timezoneId,
            };
        const detail = (0, true_solar_time_1.calculateTrueSolarTime)(civilTimeInput, {
            longitude: input.longitude,
        });
        const [dateStr, timeStr] = detail.trueSolarDateTime.split(' ');
        const [yS, mS, dS] = dateStr.split('-');
        const [hS, minS, secS] = timeStr.split(':');
        solarYear = parseInt(yS, 10);
        solarMonth = parseInt(mS, 10);
        solarDay = parseInt(dS, 10);
        solarHour = parseInt(hS, 10);
        solarMinute = parseInt(minS, 10);
        solarSecond = parseInt(secS, 10);
        solarTimeInfo = {
            trueSolarTime: detail.trueSolarTime.substring(0, 5),
            trueSolarDateTime: detail.trueSolarDateTime,
            solarDate: dateStr,
            standardMeridian: detail.standardMeridian,
            longitudeCorrectionMinutes: detail.longitudeCorrectionMinutes,
            equationOfTimeMinutes: detail.equationOfTimeMinutes,
            algorithm: detail.algorithm,
        };
    }
    // ── 3. Generate Four Pillars ─────────────────────────────────────────────
    const { pillars, eightChar, dayStem } = (0, pillars_1.generatePillarsFromSolar)(solarYear, solarMonth, solarDay, solarHour, solarMinute, solarSecond, (_e = input.dayBoundaryMode) !== null && _e !== void 0 ? _e : 'MIDNIGHT_00');
    // ── 4. Day Master ────────────────────────────────────────────────────────
    const dayMaster = (0, pillars_1.getStemInfo)(dayStem);
    // ── 5. Da Yun Cycles ─────────────────────────────────────────────────────
    const daYun = (0, cycles_1.calculateDaYun)(eightChar, input.gender, solarYear);
    // ── 6. Branch Interactions ───────────────────────────────────────────────
    const interactions = (0, interactions_1.detectInteractions)({
        year: pillars.year.branch,
        month: pillars.month.branch,
        day: pillars.day.branch,
        hour: (_g = (_f = pillars.hour) === null || _f === void 0 ? void 0 : _f.branch) !== null && _g !== void 0 ? _g : '',
    });
    const calculationSolar = createDateTime(solarYear, solarMonth, solarDay, input.hour === undefined ? undefined : solarHour, input.hour === undefined ? undefined : solarMinute, input.hour === undefined ? undefined : solarSecond);
    const lunarCalendar = createLunarDateTime(calculationSolar);
    return {
        pillars,
        dayMaster,
        daYun,
        interactions,
        solarTimeInfo,
        calendar: {
            inputType,
            civilSolar,
            calculationSolar,
            lunar: lunarCalendar.lunar,
            zodiac: lunarCalendar.zodiac,
        },
        metadata: {
            trueSolarTimeApplied: enableTST,
            dayBoundaryMode: (_h = input.dayBoundaryMode) !== null && _h !== void 0 ? _h : 'MIDNIGHT_00',
            longitude: (_j = input.longitude) !== null && _j !== void 0 ? _j : null,
            timezoneBasis: (_l = (_k = input.timezone) !== null && _k !== void 0 ? _k : input.timezoneId) !== null && _l !== void 0 ? _l : null,
            dstOffset: (_m = input.dstOffset) !== null && _m !== void 0 ? _m : 0,
        },
    };
}
